otherlibs/systhreads/st_pthreads.h

```diff
@@ -103,19 +103,26 @@ typedef struct {
   pthread_cond_t is_free;         /* signaled when free */
 } st_masterlock;
 
-static void st_masterlock_init(st_masterlock * m)
+/* Returns non-zero on failure */
+static int st_masterlock_init(st_masterlock * m)
 {
+  int rc;
   if (!m->init) {
-    // FIXME: check errors
-    pthread_mutex_init(&m->lock, NULL);
-    pthread_cond_init(&m->is_free, NULL);
+    rc = pthread_mutex_init(&m->lock, NULL);
+    if (rc != 0) goto out_err;
+    rc = pthread_cond_init(&m->is_free, NULL);
+    if (rc != 0) goto out_err2;
     m->init = 1;
   }
   m->busy = 1;
   atomic_store_rel(&m->waiters, 0);
+  return 0;
 
-  return;
-};
+ out_err2:
+  pthread_mutex_destroy(&m->lock);
+ out_err:
+  return rc;
```

Comment:
No change necessary, but noting that since we don't know which operation resulted in the non-`0` return value, all we can conclude is that `0` is success and non-`0` is an error. In particular, one cannot usefully inspect the error number.

Reply:
I think we can simplify this by thinking about which errors would be useful to ultimately signal to the mutator -- there can only really be `Out_of_memory` or another exception to signal that a thread cannot be created (to distinguish between memory exhaustion and other system limits like thread limits).

In this function, the only way the `pthread_mutex_init` or `pthread_mutex_cond` functions can fail is with a memory or resource error (`EAGAIN` or `ENOMEM`), since the functions aren't called in a way that could return `EINVAL`.  So just returning a negative error code of -1 to indicate an OOM condition would be simpler than propagating the results of one of the mutex calls failing, I think.

Reply:
It is true that you do not know which one failed, but e.g. in case of ENOMEM this does not really matter.

The docs also mention EBUSY and the function is indeed used to reinitialize a mutex after fork, so since what is done for fork+threads is not really standards-compliant, it is best to keep the explicit error message for debugging. After spending an hour trying to figure out the best course of action here, I am thinking we might be overthinking it. 

Raising an exception à la sync_check_error (which correctly maps ENOMEM to Out_of_memory) is doable, after fixing the handling of errors during domain creation.